core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -84,9 +88,102 @@ abstract class InterBrokerSendThread(name: String,
     }
   }
 
-  def wakeup(): Unit = networkClient.wakeup()
+  private def checkDisconnects(now: Long): Unit = {
+    // any disconnects affecting requests that have already been transmitted will be handled
+    // by NetworkClient, so we just need to check whether connections for any of the unsent
+    // requests have been disconnected; if they have, then we complete the corresponding future
+    // and set the disconnect flag in the ClientResponse
+    for (node <- unsentRequests.nodes.asScala) {
+      if (networkClient.connectionFailed(node)) {
+        // Remove entry before invoking request callback to avoid callbacks handling
+        // coordinator failures traversing the unsent list again.
+        val requests = unsentRequests.remove(node)
+        for (request <- requests.asScala) {
+          val handler = request.callback
+          val authenticationException = networkClient.authenticationException(node)
+          if (authenticationException == null) {
+            handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
+              handler, request.destination, now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ ,
+              null /* versionMismatch */ , null /* responseBody */))
+          } else
+            debug(s"Failed to send the following request due to authentication error: ${request.toString}")
```

Comment:
We should make sure the handler still gets invoked in this case (we can treat it the same as a disconnect). As for the authentication exception, maybe we should increase verbosity to `ERROR`?